File: ruby/space-age/space_age.rb
class SpaceAge
  
  SECONDS_IN_1YEAR_ON_EARTH = 60 * 60 * 24 * 365.25
  
  SPACE_AGES = {
    mercury: 0.2408467,
    venus: 0.61519726,
    earth: 1,
    mars: 1.8808158,
    jupiter: 11.862615,
    saturn: 29.447498,
    uranus: 84.016846,
    neptune: 164.79132
  }
  
  
  def initialize(second)
    @year_on_earth = second / SECONDS_IN_1YEAR_ON_EARTH
  end

  SPACE_AGES.each {|key, value|
    define_method "on_#{key}" do
      (@year_on_earth / value).round(2)
    end
  }
  
end

File: haskell/space-age/src/SpaceAge.hs
module SpaceAge (Planet (..), ageOn) where

data Planet
  = Mercury
  | Venus
  | Earth
  | Mars
  | Jupiter
  | Saturn
  | Uranus
  | Neptune

earthYearInSeconds :: Float
earthYearInSeconds = 31557600.0

-- 各惑星の地球年に対する公転周期比
orbitalPeriod :: Planet -> Float
orbitalPeriod Mercury = 0.2408467
orbitalPeriod Venus = 0.61519726
orbitalPeriod Earth = 1.0
orbitalPeriod Mars = 1.8808158
orbitalPeriod Jupiter = 11.862615
orbitalPeriod Saturn = 29.447498
orbitalPeriod Uranus = 84.016846
orbitalPeriod Neptune = 164.79132

ageOn :: Planet -> Float -> Float
ageOn planet seconds = seconds / earthYearInSeconds / orbitalPeriod planet
